### PR TITLE
Set up mypy and fix the issues it reported

### LINKED ISSUE #287: Add mypy for static type checking
Setup _mypy_ for static type checking

http://mypy-lang.org/
https://mypy.readthedocs.io/en/stable/

### PR DESCRIPTION
**Pull request by renatav**
## Description (e.g. "Related to ...", etc.)

- Added mypy pre-commit check. Use version 1.4.1, which is the last version that supports Python 3.7
- Addressed a variety of issues that mypy reported, mostly:
  - variable of one type being assigned a value of a different type, like when converting a string to a Path instance
  - error handling when a variable can be `None`
  - incorrect typing clues that were previously added
  - rework of roles description models, remove `DelegatedRole`, `TargetsRole` can now contain a list of delegated `TargetsRole` objects
- Some generation of keys fixes, in cases when the keystore should have a password or when it should be printed to console
- Added additional type hints

Closes #287 

Did no test cli, it's possible that this will introduce some regressions, but since the next big chunk of work will add api tests, all issues, if there are any, will be addressed then

## Code review checklist (for code reviewer to complete)

- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [x] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/taf/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: taf/constants.py [hunk 1]
```diff
@@ -1,21 +1,37 @@
...
-    "number": 1,
-    "threshold": 1,
-    "yubikey": False,
-    "scheme": DEFAULT_RSA_SIGNATURE_SCHEME,
-    "length": 3072,
-    "passwords": None,
-    "terminating": True,
-}
+
+class RoleSetupParams:
```
Comment (n-dusan) [inline taf/constants.py:12]: Could we use attrs instead of a regular class?

File: taf/git.py [hunk 2]
```diff
@@ -401,6 +419,9 @@ def branch_exists(
         a remote branch exists.
         """
         repo = self.pygit_repo
+        if repo is None:
+            # TODO default to _git
+            return False
```
Comment (n-dusan) [inline taf/git.py:424]: Do we want this code to raise `GitError`? It looks like we're raising errors on `if repo is None`, but not here?

**Review by n-dusan — CHANGES_REQUESTED** [review]
Two minor comments. This looks really good!

File: taf/git.py (hunk 2 above)
Comment (renatav) [reply taf/git.py:424]: I think the only reason is that this was the first such mypy error that I tried addressing and I initially thought that pygit2 instantiaon errors would get raised somewhere, then realized that that would not be the case and that we would return incorrect values from these functions.

**Review by n-dusan — APPROVED** [review]

**Pull request merged**